embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/execution/HttpUrlConnectionRequestExecutionService.kt

```diff
@@ -1,73 +1,62 @@
     override fun attemptHttpRequest(
         payloadStream: () -> InputStream,
         envelopeType: SupportedEnvelopeType,
-    ): ApiResponse {
+    ): ExecutionResult {
         val apiRequest = envelopeType.endpoint.getApiRequestFromEndpoint()
-
-        try {
+        var headersProvider = { emptyMap<String, String>() }
+        var failureReason: Throwable? = null
+        val responseCode = try {
             val httpUrlConnection = createUrlConnection(apiRequest)
-
             httpUrlConnection.outputStream?.use { outputStream ->
                 payloadStream().use { inputStream ->
                     inputStream.copyTo(outputStream)
                 }
             }
-
             httpUrlConnection.connect()
-
-            return handleHttpUrlConnectionResponse(httpUrlConnection, envelopeType.endpoint)
+            headersProvider = { readHttpResponseHeaders(httpUrlConnection) }
+            readResponseCode(httpUrlConnection)
         } catch (throwable: Throwable) {
-            return ApiResponse.Incomplete(throwable)
+            failureReason = throwable
+            null
         }
-    }
-
-    private fun handleHttpUrlConnectionResponse(httpUrlConnection: HttpURLConnection, endpoint: Endpoint): ApiResponse {
-        val responseCode = readResponseCode(httpUrlConnection)
-        val responseHeaders = readHttpResponseHeaders(httpUrlConnection)
 
-        return when (responseCode) {
-            HttpURLConnection.HTTP_OK -> ApiResponse.Success(null, null)
-            HttpURLConnection.HTTP_NOT_MODIFIED -> ApiResponse.NotModified
-            HttpURLConnection.HTTP_ENTITY_TOO_LARGE -> ApiResponse.PayloadTooLarge
-            TOO_MANY_REQUESTS -> ApiResponse.TooManyRequests(endpoint, responseHeaders["Retry-After"]?.toLongOrNull())
-
-            NO_HTTP_RESPONSE -> {
-                ApiResponse.Incomplete(IllegalStateException("Connection failed or unexpected response code"))
-            }
-
-            else -> ApiResponse.Failure(responseCode, responseHeaders)
-        }
+        return getResult(
+            endpoint = envelopeType.endpoint,
+            responseCode = responseCode,
+            headersProvider = headersProvider,
+            clientError = failureReason,
+        )
     }
 
-    private fun readResponseCode(httpUrlConnection: HttpURLConnection): Int = try {
+    private fun readResponseCode(httpUrlConnection: HttpURLConnection): Int? = try {
         httpUrlConnection.responseCode
     } catch (throwable: Throwable) {
-        NO_HTTP_RESPONSE
+        null
     }
 
-    private fun readHttpResponseHeaders(httpUrlConnection: HttpURLConnection): Map<String, String> {
-        return httpUrlConnection.headerFields?.mapValues { it.value.joinToString() } ?: emptyMap()
+    private fun readHttpResponseHeaders(httpUrlConnection: HttpURLConnection): Map<String, String> = try {
+        httpUrlConnection.headerFields?.mapValues { it.value.joinToString() } ?: emptyMap()
+    } catch (throwable: Throwable) {
```

Comment:
is this really needed? can headerFields throw an exception?

Reply:
I don't know what the implementation actually does, so I want to be safe. We had an issue in Unity where getting headers was returning an NPE